### ISSUE TITLE
Prediction new problem

### ISSUE THREAD

**Issue by MarioProjects** [orobix/Prototypical-Networks-for-Few-shot-Learning-PyTorch#15]
Hi guys! I am at this point and I do not just understand it, I already have the trained model. For the test, what I understand that should be done is to calculate the embedding of my sample and see which centroid is closest to classify ... If I have trained with 5000 classes I do not understand at all as in the test phase it is necessary to pass a set of support and another of query.

Taking the implementation of @ale316 , predict (support_x, support_y, query_x, query_y = None), we will pass support sets of the training set and querys of which we do not know its y (therefore we equate it to None) and ... as takes into account the 5000 classes if the support only includes 30 (to say some number)

My idea is to create all the training embeddings and then generate the centroids -> After that generate the embeddings for the test samples and get the class with the distance euclidean to the centroids... I am correct? But you dont do that,

> Sorry for the spam here, but for posterity:
> 
> I wrote a function that should return predictions given:
> 
> * a tensor `support_x` of size (n_support, 1024)
> * a tensor `support_y` of size (n_support,)
> * a tensor `query_x` of size (n_query, 1024)
> 
> ```python
> def predict(support_x, support_y, query_x, query_y=None):
>     support_x = support_x.to('cpu')
>     support_y = support_y.to('cpu')
>     query_x = query_x.to('cpu')
> 
>     classes = torch.unique(support_y)
>     n_classes = len(classes)
>     n_query = len(query_x)
> 
>     # get a list of tensors of support_y for each class
>     support_idxs = list(map(lambda c: support_y.eq(c).nonzero().squeeze(1), classes))
> 
>     # take the mean of tensors for each class to create a centroid
>     prototypes = torch.stack([support_x[idx_list].mean(0) for idx_list in support_idxs])
> 
>     # finds the euclidean distances between each query_x and each centroid
>     dists = euclidean_dist(query_x, prototypes)
> 
>     # run it through softmax
>     log_p_y = F.log_softmax(-dists, dim=1)
> 
>     # lists the idx (label) of the closest centroid for each query_x
>     _, y_hat = log_p_y.max(1)
>     labels = [classes[i] for i in y_hat.squeeze()]
> 
>     return labels
> ```
> 
> Now, I still have some doubts:
> 
> * Is this correct? **Edit**: yes it is
> * Why do we even softmax the distances, instead of just taking the min?

_Originally posted by @MarioProjects in https://github.com/orobix/Prototypical-Networks-for-Few-shot-Learning-PyTorch/issues/8#issuecomment-450552865_

**Comment by xueliang** [comment]
When I read the code, I have the same question as you.
I think that after training, it is better to save the prototypes for the further testing.

**Comment by dnlcrl** [comment]
I added a comment here, https://github.com/orobix/Prototypical-Networks-for-Few-shot-Learning-PyTorch/issues/8#issuecomment-500351221, please note that we are not the paper's writers, so it would be better to ask them (here: https://github.com/jakesnell/prototypical-networks) about those kind of datails. Closing the issue for now, feel free to re-open it if needed.
